### PR TITLE
Iaas template update (WIP)

### PR DESCRIPTION
**Pull request by ArxdSilva**
this was ment to solve tsuru/tsuru#1539

### CODE REVIEW

File: api/iaas.go [hunk 1]
```diff
@@ -225,13 +225,17 @@ func templateUpdate(w http.ResponseWriter, r *http.Request, token auth.Token) (e
 		return &errors.HTTP{Code: http.StatusBadRequest, Message: err.Error()}
 	}
 	templateName := r.URL.Query().Get(":template_name")
+	iaasName := r.Form.Get("IaaSName")
```
Comment (andrestc) [inline api/iaas.go]: You should check if the `iaasName` iaas a valid iaas name (it exists).

File: api/iaas.go [hunk 2]
```diff
@@ -225,13 +225,17 @@ func templateUpdate(w http.ResponseWriter, r *http.Request, token auth.Token) (e
...
 	dbTpl, err := iaas.FindTemplate(templateName)
 	if err != nil {
 		if err == mgo.ErrNotFound {
 			return &errors.HTTP{Code: http.StatusNotFound, Message: "template not found"}
 		}
 		return err
 	}
+	if (dbTpl.IaaSName != iaasName) && (iaasName != "") {
```
Comment (andrestc) [inline api/iaas.go]: I think you don't need the first check: `(dbTpl.IaaSName != iaasName)`.

**Comment by andrestc** [comment]
Left some commends, I think we should add some tests for this.

File: api/iaas.go (hunk 1 above)
Comment (ArxdSilva) [inline api/iaas.go]: It already does [here](https://github.com/tsuru/tsuru/blob/master/iaas/iaas.go#L115-L124) after returning `dbTpl.Update`

File: api/iaas.go [hunk 3]
```diff
@@ -148,23 +150,27 @@ func templateCreate(w http.ResponseWriter, r *http.Request, token auth.Token) (e
...
 	}
-	evt, err := event.New(&event.Opts{
-		Target:     event.Target{Type: event.TargetTypeIaas, Value: paramTemplate.IaaSName},
-		Kind:       permission.PermMachineTemplateCreate,
-		Owner:      token,
-		CustomData: event.FormToCustomData(r.Form),
-		Allowed:    event.Allowed(permission.PermMachineReadEvents, iaasCtx),
-	})
+	_, err = iaas.FindTemplate(paramTemplate.Name)
 	if err != nil {
```
Comment (andrestc) [inline api/iaas.go]: I would try to be a little more carefull here, something like:
```go
if err != nil {
    if err == mgo.ErrNotFound {
        return &errors.HTTP{Code: http.StatusConflict, Message: fmt.Sprintf("template name \"%s\" already used", paramTemplate.Name)}
    }
    return err
}
... the rest of the code here.
```

File: api/iaas.go [hunk 4]
```diff
@@ -148,23 +150,27 @@ func templateCreate(w http.ResponseWriter, r *http.Request, token auth.Token) (e
 	if !allowed {
 		return permission.ErrUnauthorized
 	}
-	evt, err := event.New(&event.Opts{
```
Comment (andrestc) [inline api/iaas.go]: I think we could create the event before checking for the template existence. I think it's good to have this failure recorded on the events.

**Comment by andrestc** [comment]
LGTM

**Pull request merged**